File: experiments-local/experiment-mu-glm52-ompfix/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep2/src/tools.ts
/**
 * Brazilian Soccer MCP Server — MCP tool layer
 * --------------------------------------------
 * Context block:
 *   Pure, side-effect-free module that defines the MCP tool schemas and a
 *   dispatcher mapping tool names + arguments to query-engine calls. Split out
 *   of index.ts so unit tests can exercise dispatch without starting the stdio
 *   transport. index.ts imports `createDispatcher`/`toolDefinitions` here and
 *   wires them to the Server.
 */

import type { Dataset } from "./loader.js";
import {
  bestRecord,
  goalStats,
  headToHead,
  queryMatches,
  queryPlayers,
  resolveTeam,
  standings,
  teamStats,
  topScoringTeams,
} from "./queries.js";

export interface ToolDef {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
}

/** The nine MCP tools exposed to the LLM. */
export function toolDefinitions(): ToolDef[] {
  return [
    {
      name: "search_matches",
      description:
        "Search Brazilian soccer match data by team, opponent, home/away team, competition, season, date range, stage or round. Returns a list of matches with date, scores, competition and round.",
      inputSchema: {
        type: "object",
        properties: {
          team: { type: "string", description: "Team name (home or away)." },
          opponent: { type: "string", description: "Opponent team name; pair with team for head-to-head match lists." },
          homeTeam: { type: "string" },
          awayTeam: { type: "string" },
          competition: { type: "string", description: "e.g. Brasileirão, Copa do Brasil, Copa Libertadores, Serie A, Serie B." },
          season: { type: "integer", description: "Year, e.g. 2023." },
          startDate: { type: "string", description: "ISO date YYYY-MM-DD." },
          endDate: { type: "string", description: "ISO date YYYY-MM-DD." },
          stage: { type: "string", description: "Tournament stage (Libertadores), e.g. final, semifinals, group stage." },
          round: { type: "string" },
          limit: { type: "integer", description: "Max results (default 50)." },
        },
      },
    },
    {
      name: "team_statistics",
      description:
        "Return aggregate statistics (wins, draws, losses, goals for/against, points) for a team, optionally filtered by season, competition and home/away.",
      inputSchema: {
        type: "object",
        properties: {
          team: { type: "string" },
          season: { type: "integer" },
          competition: { type: "string" },
          homeAway: { type: "string", enum: ["home", "away", "all"] },
        },
        required: ["team"],
      },
    },
    {
      name: "head_to_head",
      description: "Compare two teams head-to-head: wins, draws, goals across all matches (optionally a single season).",
      inputSchema: {
        type: "object",
        properties: {
          teamA: { type: "string" },
          teamB: { type: "string" },
          season: { type: "integer" },
        },
        required: ["teamA", "teamB"],
      },
    },
    {
      name: "search_players",
      description: "Search the FIFA player dataset by name, nationality, club, position and minimum overall rating. Sort by overall/potential/age/name.",
      inputSchema: {
        type: "object",
        properties: {
          name: { type: "string" },
          nationality: { type: "string", description: "e.g. Brazil." },
          club: { type: "string" },
          position: { type: "string", description: "e.g. ST, LW, GK, CDM." },
          minOverall: { type: "integer" },
          sortBy: { type: "string", enum: ["overall", "potential", "age", "name"] },
          limit: { type: "integer" },
        },
      },
    },
    {
      name: "competition_standings",
      description: "Compute a league table (standings) for a competition and season from match results. Points are 3-for-a-win; sorted by points, goal difference, goals for.",
      inputSchema: {
        type: "object",
        properties: {
          competition: { type: "string" },
          season: { type: "integer" },
        },
      },
    },
    {
      name: "goal_statistics",
      description: "Aggregate goal statistics: average goals per match, home/away/draw win rates, and biggest victories (margin >= 4).",
      inputSchema: {
        type: "object",
        properties: {
          competition: { type: "string" },
          season: { type: "integer" },
          team: { type: "string" },
        },
      },
    },
    {
      name: "best_record",
      description: "Top 5 teams by record in a competition/season, split by home or away.",
      inputSchema: {
        type: "object",
        properties: {
          competition: { type: "string" },
          season: { type: "integer" },
          homeAway: { type: "string", enum: ["home", "away"] },
        },
        required: ["homeAway"],
      },
    },
    {
      name: "top_scoring_teams",
      description: "Teams ranked by total goals scored in a competition/season.",
      inputSchema: {
        type: "object",
        properties: {
          competition: { type: "string" },
          season: { type: "integer" },
          limit: { type: "integer" },
        },
      },
    },
    {
      name: "resolve_team",
      description: "Normalise a free-text team name to its canonical key and display name. Useful to confirm a team exists before deeper queries.",
      inputSchema: {
        type: "object",
        properties: { name: { type: "string" } },
        required: ["name"],
      },
    },
  ];
}

function asString(v: unknown): string | undefined {
  return typeof v === "string" && v.length > 0 ? v : undefined;
}

function asInt(v: unknown): number | undefined {
  if (typeof v === "number") return v;
  if (typeof v === "string" && /^\d+$/.test(v)) return parseInt(v, 10);
  return undefined;
}

/** Create a dispatcher bound to a dataset. Returns JSON-serialisable results. */
export function createDispatcher(ds: Dataset) {
  return function handleTool(name: string, args: Record<string, unknown>): unknown {
    switch (name) {
      case "search_matches":
        return queryMatches(ds, {
          team: asString(args.team),
          opponent: asString(args.opponent),
          homeTeam: asString(args.homeTeam),
          awayTeam: asString(args.awayTeam),
          competition: asString(args.competition),
          season: asInt(args.season),
          startDate: asString(args.startDate),
          endDate: asString(args.endDate),
          stage: asString(args.stage),
          round: asString(args.round),
          limit: asInt(args.limit) ?? 50,
        });
      case "team_statistics":
        return teamStats(ds, {
          team: asString(args.team) ?? "",
          season: asInt(args.season),
          competition: asString(args.competition),
          homeAway: (asString(args.homeAway) as "home" | "away" | "all") ?? "all",
        });
      case "head_to_head":
        return headToHead(ds, asString(args.teamA) ?? "", asString(args.teamB) ?? "", asInt(args.season));
      case "search_players":
        return queryPlayers(ds, {
          name: asString(args.name),
          nationality: asString(args.nationality),
          club: asString(args.club),
          position: asString(args.position),
          minOverall: asInt(args.minOverall),
          sortBy: asString(args.sortBy) as "overall" | "potential" | "age" | "name" | undefined,
          limit: asInt(args.limit) ?? 20,
        });
      case "competition_standings":
        return standings(ds, {
          competition: asString(args.competition),
          season: asInt(args.season),
        });
      case "goal_statistics":
        return goalStats(ds, {
          competition: asString(args.competition),
          season: asInt(args.season),
          team: asString(args.team),
        });
      case "best_record":
        return bestRecord(ds, {
          competition: asString(args.competition),
          season: asInt(args.season),
          homeAway: (asString(args.homeAway) as "home" | "away") ?? "home",
        });
      case "top_scoring_teams":
        return topScoringTeams(ds, {
          competition: asString(args.competition),
          season: asInt(args.season),
          limit: asInt(args.limit) ?? 10,
        });
      case "resolve_team":
        return resolveTeam(ds, asString(args.name) ?? "");
      default:
        return { error: `Unknown tool: ${name}` };
    }
  };
}

File: experiments-local/experiment-mu-glm53-provider/runs/agent=oc-parasail_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep1/brazilian_soccer/tools.py
"""MCP tool registry: JSON schemas and dispatch for the query layer.

Context: this module is the bridge between the MCP protocol layer
(protocol.py) and the query functions (queries.py). Each tool is declared
with a JSON Schema for its arguments plus a handler function that receives
the parsed arguments and returns a JSON-serialisable dict. Handlers may
raise QueryError, which the protocol layer converts into an isError tool
result so the connected LLM can retry with corrected arguments.
"""

from __future__ import annotations

from . import queries
from .repository import DataRepository

_MATCHES_DESCRIPTION = """Search matches across all curated datasets.

Filter by team (either side, or force home/away with home_team/away_team),
opponent, competition (Brasileirão Serie A/B/C, Copa do Brasil, Copa
Libertadores), season, date range, Libertadores stage (e.g. "final",
"quarterfinals") or Copa do Brasil round. Set source to a CSV filename to
search the raw per-file data instead of the curated view. Examples:
"Flamengo vs Fluminense matches", "all Palmeiras matches in 2023",
"Copa do Brasil matches in May 2019"."""


def _schema(properties: dict, required: list[str] | None = None) -> dict:
    return {
        "type": "object",
        "properties": properties,
        "required": required or [],
    }


_STR = {"type": "string"}
_INT = {"type": "integer"}


def _limit_schema(default: int, maximum: int) -> dict:
    return {"type": "integer", "default": default, "minimum": 1, "maximum": maximum}


def build_tool_registry(repo: DataRepository) -> list[dict]:
    """Return the list of MCP tool descriptors bound to a repository."""

    def tool(name: str, description: str, schema: dict, handler) -> dict:
        return {
            "name": name,
            "description": description,
            "inputSchema": schema,
            "handler": handler,
        }

    return [
        tool(
            "search_matches",
            _MATCHES_DESCRIPTION,
            _schema(
                {
                    "team": _STR,
                    "opponent": _STR,
                    "home_team": _STR,
                    "away_team": _STR,
                    "competition": _STR,
                    "season": _INT,
                    "date_from": {"type": "string", "format": "date"},
                    "date_to": {"type": "string", "format": "date"},
                    "stage": _STR,
                    "round": _INT,
                    "venue": {"type": "string", "enum": ["home", "away"]},
                    "source": _STR,
                    "limit": _limit_schema(25, 200),
                    "sort": {"type": "string", "enum": ["date_desc", "date_asc"]},
                }
            ),
            lambda args: queries.search_matches(repo, **args),
        ),
        tool(
            "head_to_head",
            "Head-to-head record between two teams: wins, draws, goals and the "
            "match list. Use team spellings like 'Flamengo', 'Palmeiras-SP' or "
            "'America-MG'. Example: compare Palmeiras and Santos.",
            _schema(
                {
                    "team_a": _STR,
                    "team_b": _STR,
                    "competition": _STR,
                    "season": _INT,
                    "limit": _limit_schema(25, 200),
                },
                ["team_a", "team_b"],
            ),
            lambda args: queries.head_to_head(repo, **args),
        ),
        tool(
            "team_stats",
            "Win/draw/loss record, goals scored and conceded, home/away splits and "
            "per-competition breakdowns for one team. Examples: Corinthians home "
            "record in 2022 (team='Corinthians', season=2022, venue handled by the "
            "home section), Palmeiras record in the 2019 Brasileirão.",
            _schema(
                {
                    "team": _STR,
                    "season": _INT,
                    "competition": _STR,
                    "limit_recent": {"type": "integer", "default": 5, "minimum": 0},
                },
                ["team"],
            ),
            lambda args: queries.team_stats(repo, **args),
        ),
        tool(
            "team_rankings",
            "Rank all teams by an aggregate metric over the (optionally filtered) "
            "match set: points, wins, away_wins, away_points, goals_for, win_rate, "
            "home_win_rate, away_win_rate, etc. Example: best away record in the "
            "Brasileirão (metric='away_points').",
            _schema(
                {
                    "competition": _STR,
                    "season": _INT,
                    "metric": {
                        "type": "string",
                        "default": "points",
                        "enum": [
                            "points", "wins", "draws", "losses", "matches",
                            "goals_for", "goals_against", "goal_diff", "win_rate",
                            "home_wins", "home_points", "home_win_rate",
                            "away_wins", "away_points", "away_win_rate",
                        ],
                    },
                    "limit": {"type": "integer", "default": 10, "minimum": 1, "maximum": 100},
                }
            ),
            lambda args: queries.team_rankings(repo, **args),
        ),
        tool(
            "find_team",
            "Resolve any team spelling ('Palmeiras-SP', 'Grêmio', 'Atlético "
            "Paranaense', 'Fla') to the club entities in the datasets, with match "
            "and player counts. Use it to disambiguate names like 'America' "
            "(America-MG vs America-RN) or 'Atletico' (MG, GO, PR).",
            _schema({"query": _STR}, ["query"]),
            lambda args: queries.find_team(repo, **args),
        ),
        tool(
            "search_players",
            "Search the FIFA player database by name, nationality ('Brazil'), "
            "club, position (FIFA code like 'ST' or group like 'forward', "
            "'goalkeeper'), minimum overall rating or maximum age. Examples: all "
            "Brazilian players, highest-rated players at Grêmio, forwards from "
            "Santos.",
            _schema(
                {
                    "name": _STR,
                    "nationality": _STR,
                    "club": _STR,
                    "position": _STR,
                    "min_overall": _INT,
                    "max_age": _INT,
                    "sort": {"type": "string", "enum": ["overall", "potential", "age", "name"]},
                    "limit": _limit_schema(25, 200),
                }
            ),
            lambda args: queries.search_players(repo, **args),
        ),
        tool(
            "player_detail",
            "Full FIFA profile (ratings, skills, contract, physical attributes) "
            "for one player by name or FIFA id. Example: 'Gabriel Barbosa'.",
            _schema({"name": _STR, "id": _INT}),
            lambda args: queries.player_detail(repo, player_id=args.pop("id", None), **args),
        ),
        tool(
            "standings",
            "League table calculated from match results for one competition "
            "season, with champion and (for Serie A) the relegation zone. "
            "Examples: 2019 Brasileirão (competition='Brasileirão Serie A', "
            "season=2019), who was relegated in 2020.",
            _schema({"competition": _STR, "season": _INT}, ["competition", "season"]),
            lambda args: queries.standings(repo, **args),
        ),
        tool(
            "competition_info",
            "Describe the competitions in the datasets: seasons covered, match "
            "counts, source files. Call without arguments to list everything.",
            _schema({"competition": _STR}),
            lambda args: queries.competition_info(repo, **args) if args else queries.competition_info(repo),
        ),
        tool(
            "biggest_wins",
            "Largest victory margins in the dataset, optionally filtered by "
            "competition and season.",
            _schema(
                {
                    "competition": _STR,
                    "season": _INT,
                    "limit": {"type": "integer", "default": 10, "minimum": 1, "maximum": 100},
                }
            ),
            lambda args: queries.biggest_wins(repo, **args),
        ),
        tool(
            "stats_summary",
            "Aggregate statistics: average goals per match, home/away win rates, "
            "draws, biggest win. Filter by competition and season to compare "
            "e.g. the 2018 and 2019 Brasileirão seasons.",
            _schema({"competition": _STR, "season": _INT}),
            lambda args: queries.stats_summary(repo, **args),
        ),
        tool(
            "derby_matches",
            "Matches between traditional rivals (Fla-Flu, Grenal, Majestoso, "
            "Choque-Rei, Derby Paulista, Ba-Vi, Atletiba...). Example: all "
            "derbies in 2023.",
            _schema(
                {
                    "season": _INT,
                    "competition": _STR,
                    "limit": {"type": "integer", "default": 50, "minimum": 1, "maximum": 200},
                }
            ),
            lambda args: queries.derby_matches(repo, **args),
        ),
    ]


def call_tool(tools: list[dict], name: str, arguments: dict) -> dict:
    """Dispatch a tools/call request; raises KeyError for unknown tools."""
    for entry in tools:
        if entry["name"] == name:
            return entry["handler"](arguments or {})
    raise KeyError(name)
